x-pack/test/security_functional/tests/expired_session/basic_functionality.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import expect from '@kbn/expect';
+import { SESSION_ERROR_REASON_HEADER } from '@kbn/security-plugin/common/constants';
+import { setTimeout as setTimeoutAsync } from 'timers/promises';
+import { parse } from 'url';
+import { FtrProviderContext } from '../../ftr_provider_context';
+
+export default function ({ getService, getPageObjects }: FtrProviderContext) {
+  const browser = getService('browser');
+  const deployment = getService('deployment');
+  const PageObjects = getPageObjects(['security', 'common']);
+
+  describe('Basic functionality', function () {
+    this.tags('includeFirefox');
+
+    before(async () => {
+      await PageObjects.security.forceLogout();
+    });
+
+    afterEach(async () => {
+      // NOTE: Logout needs to happen before anything else to avoid flaky behavior
+      await PageObjects.security.forceLogout();
+    });
+
+    it('should attach msg=SESSION_EXPIRED to the redirect URL when redirecting to /login if the session has expired when trying to access a page', async () => {
+      await login();
+
+      // Kibana will log the user out automatically 5 seconds before the `xpack.security.session.idleTimeout` timeout.
+      // To simulate what will happen if this doesn't happen, navigate to a non-Kibana URL to ensure Kibana isn't running in the browser.
+      await browser.get('data:,');
+
+      // Sessions expire based on the `xpack.security.session.idleTimeout` config and is 10s in this test
+      await setTimeoutAsync(11000);
+
+      await PageObjects.common.navigateToUrl('home', '', {
+        ensureCurrentUrl: false,
+        shouldLoginIfPrompted: false,
+      });
+
+      const currentURL = parse(await browser.getCurrentUrl());
+      expect(currentURL.pathname).to.eql('/login');
+
+      expect(await PageObjects.security.loginPage.getInfoMessage()).to.be(
+        'Your session has timed out. Please log in again.'
+      );
+    });
+
+    it(`should handle returned ${SESSION_ERROR_REASON_HEADER} header when the server response is 401`, async () => {
+      // TODO: Even though the anonymous path is registered as anonymous, we're still redirected to the `/login` page. So for now I can't test the base case and have to trust that this works as intented
+      // await goToAnonymousPath();
+
+      // // First verify that without being logged in, we get the expected result if requesting a protected API
+      // const notAuthenticatedFailure = await fetchProtectedAPI();
+      // expect(notAuthenticatedFailure.statusCode).to.be(??);
+      // expect(notAuthenticatedFailure.reason).to.be(??);
```

Comment:
I tried to implement this (see diff below), but we're running into some other problems. Notably that some plugins (including the security plugin!) are requesting authenticated assets on anonymous pages. These API calls correctly return a `401` response, which is getting caught and redirected via our http response interceptor.

In order for this to work correctly:
1) You need to create a server-side route to render the "core app" anonymously within the test plugin. The diff below does this.
2) We need to eliminate these errant API calls that happen on anonymous pages. Some of this is within the security plugin (e.g. calls to `getCurrentUser` and calls to retrieve session info), and others are external, such as licensing info: https://github.com/elastic/kibana/issues/99501


Given the scope of changes required, I think leaving this as an open TODO makes sense. Can you create an issue to track this debt, and link the issue back here in the TODO comment?


```diff
diff --git a/x-pack/test/security_functional/fixtures/common/test_endpoints/public/plugin.tsx b/x-pack/test/security_functional/fixtures/common/test_endpoints/public/plugin.tsx
index 57879cd2520..18bfecdab20 100644
--- a/x-pack/test/security_functional/fixtures/common/test_endpoints/public/plugin.tsx
+++ b/x-pack/test/security_functional/fixtures/common/test_endpoints/public/plugin.tsx
@@ -19,12 +19,15 @@ export interface PluginStartDependencies {
 export class TestEndpointsPlugin implements Plugin<void, void, object, PluginStartDependencies> {
   public setup(core: CoreSetup<PluginStartDependencies>) {
     // Prevent auto-logout on server `401` errors.
-    core.http.anonymousPaths.register('/app/expired_session_test');
+    core.http.anonymousPaths.register('/expired_session_test');
     core.application.register({
       id: 'expired_session_test',
       title: 'Expired Session Test',
+      appRoute: '/expired_session_test',
+      chromeless: true,
       async mount({ element }) {
         (window as any).kibanaFetch = core.http.fetch;
+        element.innerText = 'Expired session test page';
         return () => ReactDOM.unmountComponentAtNode(element);
       },
     });
diff --git a/x-pack/test/security_functional/fixtures/common/test_endpoints/server/init_routes.ts b/x-pack/test/security_functional/fixtures/common/test_endpoints/server/init_routes.ts
index 16db5c82429..3b4e76a0c2e 100644
--- a/x-pack/test/security_functional/fixtures/common/test_endpoints/server/init_routes.ts
+++ b/x-pack/test/security_functional/fixtures/common/test_endpoints/server/init_routes.ts
@@ -26,6 +26,17 @@ export function initRoutes(initializerContext: PluginInitializerContext, core: C
     }
   );
 
+  core.http.resources.register(
+    {
+      path: '/expired_session_test',
+      validate: false,
+      options: { authRequired: false },
+    },
+    async (context, request, response) => {
+      return response.renderAnonymousCoreApp();
+    }
+  );
+
   const router = core.http.createRouter();
   router.post(
     {
diff --git a/x-pack/test/security_functional/tests/expired_session/basic_functionality.ts b/x-pack/test/security_functional/tests/expired_session/basic_functionality.ts
index 213079f3b85..f4e45c9db6c 100644
--- a/x-pack/test/security_functional/tests/expired_session/basic_functionality.ts
+++ b/x-pack/test/security_functional/tests/expired_session/basic_functionality.ts
@@ -52,13 +52,12 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
     });
 
     it(`should handle returned ${SESSION_ERROR_REASON_HEADER} header when the server response is 401`, async () => {
-      // TODO: Even though the anonymous path is registered as anonymous, we're still redirected to the `/login` page. So for now I can't test the base case and have to trust that this works as intented
-      // await goToAnonymousPath();
+      await goToAnonymousPath();
 
       // // First verify that without being logged in, we get the expected result if requesting a protected API
-      // const notAuthenticatedFailure = await fetchProtectedAPI();
-      // expect(notAuthenticatedFailure.statusCode).to.be(??);
-      // expect(notAuthenticatedFailure.reason).to.be(??);
+      const notAuthenticatedFailure = await fetchProtectedAPI();
+      expect(notAuthenticatedFailure.statusCode).to.be(200);
+      expect(notAuthenticatedFailure.reason).to.be('');
 
       await login();
 
@@ -85,9 +84,9 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
       // timer which normally would log the user out 5 seconds before their
       // session expires, isn't scheduled.
       async function goToAnonymousPath() {
-        await browser.get(`${deployment.getHostPort()}/app/expired_session_test`);
+        await browser.get(`${deployment.getHostPort()}/expired_session_test`);
         const currentURL = parse(await browser.getCurrentUrl());
-        expect(currentURL.pathname).to.eql('/app/expired_session_test');
+        expect(currentURL.pathname).to.eql('/expired_session_test');
       }
 
       async function fetchProtectedAPI() {
@@ -95,7 +94,7 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
         // logged out and redirected to `/login`. If we're not on this page,
         // getting `window.kibanaFetch` later on will not work.
         const currentURL = parse(await browser.getCurrentUrl());
-        expect(currentURL.pathname).to.eql('/app/expired_session_test');
+        expect(currentURL.pathname).to.eql('/expired_session_test');
 
         // Exectue an AJAX request from within the browser to the Kibana server
         // and return the result
```